8264774: Implementation of Foreign Function and Memory API (Incubator)

mcimadamore opened the pull request:
This PR contains the API and implementation changes for JEP-412 [1]. A more detailed description of such changes, to avoid repetitions during the review process, is included as a separate comment.

[1] - https://openjdk.java.net/jeps/412

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264774](https://bugs.openjdk.java.net/browse/JDK-8264774): Implementation of Foreign Function and Memory API (Incubator)


### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to 7545f71f91def59df88fd71289eda1bbba3fb72e
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to a80d8180417e875879d19574545b598a07c57a49
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 8de9da36fabf12814d7e3b0dbf62f50be5acb021
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to 352c287f4810d3201b4574e05f316853fbbe85aa


### Contributors
 * Paul Sandoz `<psandoz@openjdk.org>`
 * Jorn Vernee `<jvernee@openjdk.org>`
 * Vladimir Ivanov `<vlivanov@openjdk.org>`
 * Athijegannathan Sundararajan `<sundar@openjdk.org>`
 * Chris Hegarty `<chegar@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3699/head:pull/3699` \
`$ git checkout pull/3699`

Update a local copy of the PR: \
`$ git checkout pull/3699` \
`$ git pull https://git.openjdk.java.net/jdk pull/3699/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3699`

View PR using the GUI difftool: \
`$ git pr show -t 3699`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3699.diff">https://git.openjdk.java.net/jdk/pull/3699.diff</a>

</details>

mcimadamore commented:
Here we list the main changes introduced in this PR.

### Managing memory segments: `ResourceScope`

This PR introduces a new abstraction, namely `ResourceScope` which is used to manage the lifecycle of resources associated with off-heap memory (such as `MemorySegment`, `VaList`, etc). This is probably the biggest change in the API, as `MemorySegment` is no longer `AutoCloseable`: it instead features a *scope accessor* which can be used to access the memory segment's `ResourceScope`; the `ResourceScope` is the new  `AutoCloseable`. In other words, code like this:

```
try (MemorySegment segment = MemorySegment.allocateNative(100)) {
   ...
}   
```

Now becomes like this:

```
try (ResourceScope scope = ResourceScope.ofConfinedScope()) {
   MemorySegment segment = MemorySegment.allocateNative(100, scope);
   ...
}   
```

While simple cases where only one segment is allocated become a little more verbose, this new API idiom obviously scales much better when multiple segments are created with the same lifecycle. Another important fact, which is captured by the name of the `ResourceScope` factory in the above snippet, is that segments no longer feature dynamic ownership changes. These were cool, but ultimately too expensive to support in the shared case. Instead, the API now requires clients to make a choice upfront (confined, shared or *implicit* - where the latter means GC-managed, like direct buffers).

### IO operation on shared buffer views

In the previous iteration of the Memory Access API we have introduced the concept of *shared* segments. Shared segments are as easy to use as confined ones, and they are as fast. One problem with shared segments was that it wasn't clear how to support IO operations on byte buffers derived from such segments: since the segment memory could be released at any time, there was simply no way to guarantee that a shared segment could not be closed in the middle of a (possibly async) IO operation.

In this iteration, shared segments are just segments backed by a *shared resource scope*. The new API introduces way to manage the new complexity, in the form of two methods `ResourceScope::acquire` and `ResourceScope::release`, respectively, which can be used to *acquire* a resource scope. When a resource scope is in the acquired state, it cannot be closed (you can think of it as some slightly better and asymmetric form of an atomic reference counter).

This means we are now finally in a position to add support for IO operations on all byte buffers, including those derived from shared segments. A big thank to @ChrisHegarty who lead the effort here. More info on this work are included in his writeup at [1].

### Allocating at speed: `SegmentAllocator`

Another abstraction introduced in this JEP is that of `SegmentAllocator`. A segment allocator is a functional interface which can be used to tell other APIs (and, crucially, the `CLinker` API) how segments should be allocated, if the need arise. For instance, think about some code which turns a Java string into a C string. Such code will invariably:

1. allocate a memory segment off heap
2. bulk copy (where possible) the content of the Java string into the off-heap segment
3. add a NULL terminator

So, in (1) such string conversion routine need to allocate a new off-heap segment; how is that done? Is that a call to malloc? Or something else? In the previous iteration of the Foreign Linker API, the method `CLinker::toCString` had two overloads: a simple version, only taking a Java string parameter; and a more advanced version taking a `NativeScope`. A `NativeScope` was, at its core, a custom segment allocator - but the allocation scheme was fixed in `NativeScope` as that class always acted as an arena-style allocator.

`SegmentAllocator` is like `NativeScope` in spirit, in that it helps programs allocating segments - but it does so in a more general way than `NativeScope`, since a `SegmentAllocator` is not tied to any specific allocation strategy: in fact the strategy is left there to be defined by the user. As before, `SegmentAllocator` does provide some common factories, e.g. to create an arena allocator similar to `NativeScope` - but the `CLinker` API is now free to work with _any_ implementations of the `SegmentAllocator` interface. This generalization is crucial, given that, when operating with off-heap memory, allocation performance is often the bottleneck.

Not only is `SegmentAllocator` accepted by all methods in the `CLinker` API which need to allocate memory: even the behavior of downcall method handle can now be affected by segment allocators: when linking a native function which returns a struct by value, the `CLinker` API will in fact need to dynamically allocate a segment to hold the result. In such cases, the method handle generated by `CLinker` will now accept an additional *prefix* parameter of type `SegmentAllocator` which tells `CLinker` *how* should memory be allocated for the result value.

### `MemoryAddress` as scoped entities

A natural consequence of introducing the `ResourceScope` abstraction is that now not only `MemorySegment` are associated with a scope, but even instances of `MemoryAddress` can be. This means extra safety, because passing addresses which are associated with a closed scope to a native function will issue an exception. As before, it is possible to have memory addresses which the runtime knows nothing about (those returned by native calls, or those created via `MemoryAddress::ofLong`); these addresses are simply associated with the so called *global scope* - meaning that they are not actively managed by the user and are considered to be "always alive" by the runtime (as before).

### *Virtual* downcall method handles

There are cases where the address of a downcall handle cannot be specified when a downcall method handle is linked, but can only be known subsequently, by doing more native calls. To better support these use cases, `CLinker` now provides a factory for downcall method handles which does *not* require any function entry point. Instead, such entry point will be provided *dynamically*, via an additional prefix parameter (of type `MemoryAddress`). Many thanks to @JornVernee who implemented this improvement.

### Optimized upcall support

The previous iteration of the Foreign Linker API supported intrinsification of downcall handles, which allows calls to downcall method handles to perform as fast as a regular JNI call. The dual case, calling Java code from native code (upcalls) was left unoptimized. In this iteration, @JornVernee has added intrinsics support for upcalls as well as downcalls, based on some prior work from Vlad. As for downcalls, a lot of the adaptation now happens in Java code, before we jump into the target method handle. As for the code which calls such target handle, changes have been made so that the native code can jump to the optimized entry point (if one exists) for such method handle more directly. The performance improvements with this new approach are rather nice, with `CLinker` upcalls being 3x-4x faster compared with regular upcalls via JNI.

### Accessing restricted methods

It is still the case that some of the methods in the API are "restricted" and access to these methods is disabled by default. In previous iterations, access to such methods was granted by setting a JDK read-only runtime property: `-Dforeign.restricted=permit`. In this iteration we have refined the story for accessing restricted methods, by introducing a new experimental command line option in the Java launcher, namely `--enable-native-access=<module list>`. This options accepts a list of modules (separated by commas), where a module name can also be `ALL-UNNAMED` (for the unnamed module). Adding this command line flag to the launcher has the effect of allowing access to restricted methods to a given set of modules (the list of modules specified in the command line option). Access to restricted methods from any other module not in the list is disallowed and will result in an `IllegalAccessException`.

mcimadamore commented:
Javadoc: http://cr.openjdk.java.net/~mcimadamore/JEP-412/v1/javadoc/jdk/incubator/foreign/package-summary.html
Specdiff: http://cr.openjdk.java.net/~mcimadamore/JEP-412/v1/specdiff/overview-summary.html

mcimadamore commented:
/contributor add @PaulSandoz 
/contributor add @JornVernee 
/contributor add @iwanowww
/contributor add @sundararajana
/contributor add @ChrisHegarty

mcimadamore commented:
/csr

mcimadamore commented:
/contributor add @JornVernee

mcimadamore commented:
/contributor add @iwanowww

mcimadamore commented:
/contributor add @sundararajana

mcimadamore commented:
/contributor add @ChrisHegarty

AlanBateman commented on src/java.base/share/classes/jdk/internal/module/ModuleBootstrap.java:
```diff
@@ -268,7 +268,8 @@ private static ModuleLayer boot2() {
         BootLoader.loadModule(base);
-        Modules.defineModule(null, base.descriptor(), baseUri);
+        SharedSecrets.getJavaLangAccess()
+                .addEnableNativeAccess(Modules.defineModule(null, base.descriptor(), baseUri));
```
This would be cleaner if you replace it with:

Module baseModule = Modules.defineModule(null, base.descriptor(), baseUri);
JLA.addEnableNativeAccess(baseModule);

You can use JLA in addEnableNativeAccess too, no need for "jla".

mcimadamore replied:
ok

AlanBateman commented on src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java:
```diff
@@ -378,4 +378,8 @@
+
+    Module addEnableNativeAccess(Module m);
+
+    boolean isEnableNativeAccess(Module m);
```
Probably better to co-locate these with the other module methods, and add a comment so that it's consistent with the existing methods.

mcimadamore replied:
sure I'll fix this

AlanBateman commented on src/java.base/share/classes/jdk/internal/module/ModuleBootstrap.java:
```diff
@@ -869,6 +873,21 @@ private static void addIllegalAccess(String illegalAccess,
         builder.complete();
     }
 
+    private static void addEnableNativeAccess(ModuleLayer layer) {
```
It would be useful to add a method description in the same style as the existing methods, just to keep these methods consistent.

mcimadamore replied:
and this

mcimadamore replied:
I'll fix

AlanBateman commented on src/java.base/share/classes/sun/nio/ch/IOUtil.java:
```diff
@@ -47,15 +50,30 @@ static int write(FileDescriptor fd, ByteBuffer src, long position,
+    }
+
+    static int write(FileDescriptor fd, ByteBuffer src, long position,
+                     boolean async, NativeDispatcher nd)
```
UnixAsynchronousSocketChannel is an outlier user of IOUtil and I think we need to see if we can avoid extending this class to async if possible.

AlanBateman commented on src/java.base/share/classes/sun/nio/ch/IOUtil.java:
```diff
@@ -384,6 +463,83 @@ static long read(FileDescriptor fd, ByteBuffer[] bufs, int offset, int length,
         }
     }
 
+    private static final JavaNioAccess NIO_ACCESS = SharedSecrets.getJavaNioAccess();
```
It might be cleaner to move to acquire/release methods to their own supporting class as it's not really IOUtil.

ChrisHegarty replied:
I went back and forth on this a number of times already. I think where we landed is a reasonable place, given the current shape of the code.

Scope is a private property of Buffer, and as such should be consulted anywhere where a buffer's address is being accessed. In fact, a prior prototype would not allow access to the underlying address value without the caller passing a valid handle for the buffer view's scope. It's hard to find the sweet-spot here between code reuse and safety, but the high-order bit is that the code accessing the address is auditable and testable to avoid accessing memory unsafely. Maybe there is a better alternative implementation code structure (at the cost of some duplication), but it is not obvious to me what that is (and I have given it some thought). Suggestions welcome.

Note, there is a little more follow-on work to be done in this area, if we are to expand support to other non-TCP channel implementations. Maybe investigation into possible code refactorings could be done as part of that?

AlanBateman replied:
Can you create a follow-on issue to re-visit the changes to IOUtil? The changes in this area that are in this PR will need to re-worked so that it more cleanly separate the synchronous and asynchronous usages.

AlanBateman commented on src/java.base/share/classes/java/lang/Module.java:
```diff
@@ -246,6 +250,20 @@ public ModuleLayer getLayer() {
         return null;
     }
 
+    boolean isEnableNativeAccess() {
```
Would it be possible isEnableNativeAccess and addNEnableNativeAccess to keep them consistent with the existing methods in this file.

mcimadamore replied:
Change as in "adding javadoc" ? Or is the naming of the method something that's bothering?

PaulSandoz reviewed (approved):
The Java changes are mostly familiar to me, having reviewed many PRs in the Panama repository, hence the lack of specific comments here. I did a pass through the changes and nothing major stood out to me.

The new API changes make it much easy to build APIs around memory segments, exposing and reusing resource scope, acquiring/releasing on resource scope, where necessary.

ChrisHegarty commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java:
```diff
@@ -0,0 +1,289 @@
+     * resource scope; an explicit resource scope cannot be {@link #close() closed}
+     * until all the resource scope handles acquired from it have been {@link #release(Handle)} released}.
+     * @return a resource scope handle.
+     */
```
Given recent changes, it might be good to generalise the opening sentence here. Maybe :
 " Acquires a resource scope handle associated with this resource scope. If the resource scope is explicit ... "   Or some such.

mcimadamore replied:
I'm afraid I don't get this comment ;-)

ChrisHegarty replied:
The spec for the acquire method talks only of explicit resource scopes. The comment is suggesting that the spec could be generalised a little, since it is possible to acquire a resource scope handle associated with implicit scopes.

ChrisHegarty commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java:
```diff
@@ -0,0 +1,289 @@
+   <critical operation on segment>
+} finally {
+   segment.scope().release(segmentHandle);
+}
```
I do like the symmetry in this example, but just to add an alternative idiom:
    `segmentHandle.scope().release(segmentHandle)`
, which guarantees to avoid release throwing and IAE

mcimadamore replied:
I see what you mean - I don't think I want to encourage that style too much by giving it prominence in the javadoc. It's true you can go back to the scope from the handle, so that you are guaranteed to release the right thing, but I think that should be unnecessary in most idiomatic uses.

ChrisHegarty commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java:
```diff
@@ -693,165 +581,175 @@ default MemorySegment asSlice(MemoryAddress newBase) {
      */
-    static MemorySegment allocateNative(MemoryLayout layout) {
+    static MemorySegment allocateNative(MemoryLayout layout, ResourceScope scope) {
+        Objects.requireNonNull(scope);
```
Should the allocateNative methods declare that they throw ISE, if the given ResourceScope is not alive?   ( I found myself asking this q, then considering the behaviour of a SegmentAllocator that is asked to allocate after a RS has been closed )

mcimadamore replied:
Good point, yes it should

ChrisHegarty reviewed (approved):
Like Paul, I tracked the changes to this API in the Panama repo. Most of my remaining comments are small and come from re-reading the API docs.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -205,63 +267,65 @@ static CLinker getInstance() {
     /**
-     * Converts a Java string into a null-terminated C string, using the
-     * platform's default charset, storing the result into a new native memory segment.
+     * Converts a Java string into a null-terminated C string, using the platform's default charset,
```
I believe here the goal is to be consistent with `String::getBytes`:

```
/**
     * Encodes this {@code String} into a sequence of bytes using the
     * platform's default charset, storing the result into a new byte array.
     *
     * <p> The behavior of this method when this string cannot be encoded in
     * the default charset is unspecified.  The {@link
     * java.nio.charset.CharsetEncoder} class should be used when more control
     * over the encoding process is required.
     *
     * @return  The resultant byte array
     *
     * @since      1.1
     */
    public byte[] getBytes() {
        return encode(Charset.defaultCharset(), coder(), value);
    }
```

So, you are right in that there's an element of platform-dependency here - but I think it's a dependency that users learned to "ignore" mostly. If developers want to be precise, and platform independent, they can always use the Charset-accepting method. Of course this could be revisited, but I think there is some consistency in what the API is trying to do. If, in the future, defaultCharset will just be Utf8 - well, that's ok too - as long as the method is specified to be "defaultCharset" dependent, what's behind "defaultCharset" is an implementation detail that the user will have to be aware of one way or another.

AlanBateman replied:
Naoto is working on a couple of changes in advance of JEP 400. One of these is to expose a system property with the platform encoding and I suspect that the CLinker method will want to use that instead of Charset.defaultCharset.

dfuch commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:
```diff
@@ -197,39 +199,38 @@ static int intCompare(MemoryAddress addr1, MemoryAddress addr2) {
  * <h2>Restricted methods</h2>
  * Some methods in this package are considered <em>restricted</em>. Restricted methods are typically used to bind native
  * foreign data and/or functions to first-class Java API elements which can then be used directly by client. For instance
- * the restricted method {@link jdk.incubator.foreign.MemoryAddress#asSegmentRestricted(long)} can be used to create
+ * the restricted method {@link jdk.incubator.foreign.MemoryAddress#asSegment(long, ResourceScope)} can be used to create
```
typo: `used directly by client.` => `used directly by clients.` ?

mlchung commented:
I reviewed the `--enable-native-access` related change that looks fine.

> Access to restricted methods from any other module not in the list is disallowed and will result in an IllegalAccessException.

I think you meant to say `IllegalCallerException` instead of `IllegalAccessException`.  Also do you intend to have javadoc to generate `@throw IllegalCallerException` for  the restricted methods automatically besides the javadoc description?

Making the restricted methods as `@CallerSensitive` in order to get the caller class for native access check is the proper approach.   However, some interface methods are restricted methods such as `CLinker::downcallHandle` whose the implementation method is `@CallerSensitive`.    I concern with the security issue with method handle and type aliasing.   On the other hand, `CLinker` is a sealed interface and only implemented by the platform and so it's less of a concern.   I think the interface method should also be `@CallerSensitive` so that for example a method handle for `CLinker::downcallHandle` will be produced with the proper caller-sensitive context.

dfuch commented on src/java.base/share/classes/jdk/internal/module/IllegalNativeAccessChecker.java:
```diff
@@ -0,0 +1,100 @@
+        String prefix = "jdk.module.enable.native.access.";
+        int index = 0;
+        // the system property is removed after decoding
+        String value = getAndRemoveProperty(prefix + index);
```
I am not sure what is going on with the removal of the properties, but if I'm not mistaken this is racy: from the implementation of the checker() method above, it looks as if two different threads could trigger a call to the decode() function concurrently, which can result in a random partitioning of the properties against the two checkers being instantiated, with one of them being eventually set as the system-wide checker.

mcimadamore replied:
I think the method is called during module bootstrap - I don't think there is a race in practice. This method is also called in other parts of ModuleBootstrap. The code you allude to is called during initialization of the IllegalNativeAccessChecker singleton, which should happen only once, and only from one thread.

dfuch replied:
I'll take your word for it - the use of a volatile variable to store the singleton instance made this suspicious.

mcimadamore replied:
good point - I think that came in when I adapted the code from IllegalAccessLogger - which can indeed be accessed from multiple threads. In this case the initialization is effectively part of bootstrap, and volatile is not required.

mcimadamore commented:
> I reviewed the `--enable-native-access` related change that looks fine.
> 
> > Access to restricted methods from any other module not in the list is disallowed and will result in an IllegalAccessException.
> 
> I think you meant to say `IllegalCallerException` instead of `IllegalAccessException`. Also do you intend to have javadoc to generate `@throw IllegalCallerException` for the restricted methods automatically besides the javadoc description?
> 

IllegalCalller is probably better yes - we started off with an access-like check, so things have evolved a bit. I'll also add the @throws.

> Making the restricted methods as `@CallerSensitive` in order to get the caller class for native access check is the proper approach. However, some interface methods are restricted methods such as `CLinker::downcallHandle` whose the implementation method is `@CallerSensitive`. I concern with the security issue with method handle and type aliasing. On the other hand, `CLinker` is a sealed interface and only implemented by the platform and so it's less of a concern. I think the interface method should also be `@CallerSensitive` so that for example a method handle for `CLinker::downcallHandle` will be produced with the proper caller-sensitive context.

I believe that we had to move @CallerSensitive out of interfaces because there was a test that was checking that @CS was not put on "virtual" methods.

mcimadamore commented on src/java.base/share/classes/jdk/internal/module/IllegalNativeAccessChecker.java:
```diff
@@ -0,0 +1,100 @@
+    private final boolean allowAllUnnamedModules;
+
+    private IllegalNativeAccessChecker(Set<String> allowedModuleNames, boolean allowAllUnnamedModules) {
+        this.allowedModuleNames = Collections.unmodifiableSet(allowedModuleNames);
```
Honestly, I'm not even sure why we should be concerned about mutation of the set here - since we create this with a bunch of values read from a system property... e.g. should we just store `allowedModuleNames` period?

mlchung commented:
> I believe that we had to move @CallerSensitive out of interfaces because there was a test that was checking that @cs was not put on "virtual" methods.

Good if we do have such test.    We need to re-examine this with the security team.   I suggest to create a JBS issue and follow up separately.

mcimadamore commented:
> > I believe that we had to move @CallerSensitive out of interfaces because there was a test that was checking that @cs was not put on "virtual" methods.
> 
> Good if we do have such test. We need to re-examine this with the security team. I suggest to create a JBS issue and follow up separately.

The test in question is:

test/jdk/jdk/internal/reflect/CallerSensitive/CheckCSMs.java

mlchung commented:
> test/jdk/jdk/internal/reflect/CallerSensitive/CheckCSMs.java

Yes that's the test.  That test misses to catch your case where aliasing may be possible.

dfuch commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryAddress.java:
```diff
@@ -189,12 +205,14 @@ default MemorySegment asSegmentRestricted(long bytesSize) {
     /**
-     * The off-heap memory address instance modelling the {@code NULL} address.
+     * The native memory address instance modelling the {@code NULL} address, associated
+     * with the {@link ResourceScope#globalScope() global} resource scope.
```
`{@linkplain }` ?

mcimadamore replied:
thanks - I'll do a pass - I think I probably got all of them wrong

mcimadamore commented:
> > test/jdk/jdk/internal/reflect/CallerSensitive/CheckCSMs.java
> 
> Yes that's the test. That test misses to catch your case where aliasing may be possible.



> Yes that's the test. That test misses to catch your case where aliasing may be possible.

To be clear - by aliasing you mean when the @CallerSensitive implementation is called with invokeinterface - so, e.g. doing `MethodHandles.lookup().findVirtual(CLinker.class, ...)` right?

mlchung commented:
> To be clear - by aliasing you mean when the @CallerSensitive implementation is called with invokeinterface - so, e.g. doing `MethodHandles.lookup().findVirtual(CLinker.class, ...)` right?

Yes.   The caller would be java.base if it's invoked via method handle.

mcimadamore commented:
> > To be clear - by aliasing you mean when the @CallerSensitive implementation is called with invokeinterface - so, e.g. doing `MethodHandles.lookup().findVirtual(CLinker.class, ...)` right?
> 
> Yes. The caller would be java.base if it's invoked via method handle.

I just did a test:

```
public class TestLookup {
    public static void main(String[] args) throws Throwable {
        MethodHandle handle = MethodHandles.lookup().findVirtual(CLinker.class, "downcallHandle", MethodType.methodType(MethodHandle.class, Addressable.class, MethodType.class, FunctionDescriptor.class));
        CLinker linker = CLinker.getInstance();
        handle.invoke(linker, MemoryAddress.NULL, MethodType.methodType(void.class), FunctionDescriptor.ofVoid());
    }
}
```

this fails as expected when the handle is invoked. To test I had to disable the check on CLinker.getInstance - otherwise that would have always throw anyway.

Also, on IllegalCaller vs. IllegalAccess - looking more, I think our impl throws IllegalCaller - but I believe the right exception should be IllegalAccessException, given the name of the option ("enableNativeAccess") ?

mlchung commented:
> I just did a test:
> 
> ```
> public class TestLookup {
>     public static void main(String[] args) throws Throwable {
>         MethodHandle handle = MethodHandles.lookup().findVirtual(CLinker.class, "downcallHandle", MethodType.methodType(MethodHandle.class, Addressable.class, MethodType.class, FunctionDescriptor.class));
>         CLinker linker = CLinker.getInstance();
>         handle.invoke(linker, MemoryAddress.NULL, MethodType.methodType(void.class), FunctionDescriptor.ofVoid());
>     }
> }
> ```
> 
> this fails as expected when the handle is invoked. To test I had to disable the check on CLinker.getInstance - otherwise that would have always throw anyway.

My statement was overly simplified.   If `handle` is invoked in another module B and invoked by a class in module B,  which module (the `lookup`'s module or ) do you expect be the caller to check against for native access check?    `CLinker::downcallHandle` is not caller-sensitive but its implementation is.

The method handle of a caller-sensitive method behaves as if it were called from an instruction contained in the lookup class [1].   

[1] https://download.java.net/java/early_access/jdk17/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#callsens

> Also, on IllegalCaller vs. IllegalAccess - looking more, I think our impl throws IllegalCaller - now that was done because IllegalAccess is a checked exception and we don't want a checked exception here - but the option is called "enableNativeAccess" - is that still ok?

Yes the implementation throws `IllegalCallerException` which is why I point out this.   Hmm... this seems more of `IllegalAccess` as the caller does not have access to this restricted method.  OTOH, `Module::addOpens` grants deep reflection access to the named module if the caller has access.  Otherwise, `IllegalCallerException` is thrown.  So I think it's okay to throw ICE.  Others may have a different opinion.

mcimadamore commented:
> My statement was overly simplified, sorry. If `handle` is invoked in another module B and invoked by a class in module B, which module (the `lookup`'s module or ) do you expect be the caller to check against for native access check? `CLinker::downcallHandle` is not caller-sensitive but its implementation is.
> 
> The method handle of a caller-sensitive method behaves as if it were called from an instruction contained in the lookup class [1].

I think I expect that, with caller sensitive, it is possible from a client in an "enabled" module to obtain a MethodHandle, and then pass it to an unprivileged module, which then calls it, and works ok. This matches my expectation (this trick could be used also to give a client access to an otherwise inaccessible - in the classic sense - method).

So, when working with method handles, I'm completely fine working with the lookup class as context (and not with the real caller class). 

Given all this, do you agree that current checks are not affected by interface-ness of some of the methods involved?

mcimadamore commented:
Here's a new round of javadoc/specdiff with latest changes
Javadoc: http://cr.openjdk.java.net/~mcimadamore/JEP-412/v2/javadoc/jdk/incubator/foreign/package-summary.html
Specdiff: http://cr.openjdk.java.net/~mcimadamore/JEP-412/v2/specdiff/overview-summary.html

mlchung commented:
> I think I expect that, with caller sensitive, it is possible from a client in an "enabled" module to obtain a MethodHandle, and then pass it to an unprivileged module, which then calls it, and works ok. This matches my expectation (this trick could be used also to give a client access to an otherwise inaccessible - in the classic sense - method).
> 
> So, when working with method handles, I'm completely fine working with the lookup class as context (and not with the real caller class).

Good, this matches the behavior of method handles for CSMs.   However, I think the implementation does the opposite since we don't support interface method being CSM.  Can you test that out?

> Given all this, do you agree that current checks are not affected by interface-ness of some of the methods involved?

I think the implementation does not support that.   I will also need to look into how this impacts JDK-8266010.  As I suggest earlier, I'm fine to do this as a follow up after integration.

mcimadamore commented:
> I think the implementation does not support that. I will also need to look into how this impacts JDK-8266010. As I suggest earlier, I'm fine to do this as a follow up after integration.

I've added @CS in the interface methods too. I've also added a stronger test which creates method handles in one module (which doesn't have native access) and then calls them from another module (which does NOT have native access enabled), and make sure that all call fails as expected (e.g. that caller context is that of the lookup module).

Surprisingly, CheckCSMs does NOT fail if both interface and implementation methods are @CS annotated - but if only interface is annotated, the test fails. This seems odd - since I can't find any logic in the test which check for overriding. Is the test accidentally passing?

mlchung commented:
> I've added `@CS` in the interface methods too. I've also added a stronger test which creates method handles in one module (which doesn't have native access) and then calls them from another module (which does NOT have native access enabled), and make sure that all call fails as expected (e.g. that caller context is that of the lookup module).

Thanks.
 
> Surprisingly, CheckCSMs does NOT fail if both interface and implementation methods are `@CS` annotated - but if only interface is annotated, the test fails. This seems odd - since I can't find any logic in the test which check for overriding. Is the test accidentally passing?

I create JDK-8266383 and I will look into that.

AlanBateman commented on src/java.base/share/classes/jdk/internal/module/IllegalNativeAccessChecker.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public final class IllegalNativeAccessChecker {
```
Are you sure about the name of the this class? It doesn't do any checking and it's not concerned with "illegal native access" either, instead it just provides access to the names of modules that have been granted native access.

mcimadamore replied:
I've decided to drop this class and move the logic in ModuleBootstrap. To handle unnamed modules, I had to piggy back on the special static ALL_UNNAMED module and set the flag there when decoding the command line option. This seems to be consistent with what is done in other areas.

AlanBateman replied:
Good, this looks much simpler now.

AlanBateman commented on src/java.base/share/classes/jdk/internal/module/IllegalNativeAccessChecker.java:
```diff
@@ -0,0 +1,103 @@
+    // system-wide IllegalNativeAccessChecker
+    private static IllegalNativeAccessChecker checker;
+
+    static Collection<String> enableNativeAccessModules() {
```
I assume this can be changed to Iterable<String> as you don't want anything outside of this class changing the collection.

AlanBateman commented on src/java.base/share/classes/jdk/internal/module/ModuleBootstrap.java:
```diff
@@ -869,6 +874,23 @@ private static void addIllegalAccess(String illegalAccess,
+                SharedSecrets.getJavaLangAccess().addEnableNativeAccess(module.get());
+            } else {
+                // silently skip.
+                // warnUnknownModule(ENABLE_NATIVE_ACCESS, name);
```
Maybe for later but the other options do emit a warning when unknown module is specified. If the decoding of this command line option is moved to ModuleBootstrap then most of this class will go away and you will be able to use warnUnknownModule.

AlanBateman commented on src/java.base/share/classes/java/lang/Module.java:
```diff
@@ -111,6 +111,8 @@
     private final ModuleDescriptor descriptor;
 
+    // is this module a native module
+    private volatile boolean enableNativeAccess = false;
```
Can you drop "= false", it's not needed and I don't think we want a volatile-write here.
Also the comment may date from a previous iteration as there isn't any concept of a "native module".

AlanBateman commented:
Just to double, there is no way to enable native access for modules in module layers (other than the boot layer), right?

mcimadamore commented:
> Just to

mcimadamore commented:
> Just to double, there is no way to enable native access for modules in module layers (other than the boot layer), right?

No, at the moment it is not possible to enable native access programmatically. We will explore something along those lines in the future.

AlanBateman commented on src/java.base/share/classes/jdk/internal/reflect/Reflection.java:
```diff
@@ -106,6 +106,13 @@ public static void ensureMemberAccess(Class<?> currentClass,
+    public static void ensureNativeAccess(Class<?> currentClass) {
+        Module module = currentClass.getModule();
+        if (!SharedSecrets.getJavaLangAccess().isEnableNativeAccess(module)) {
+            throw new IllegalCallerException("Illegal native access from module: " + module);
```
You may want to drop "module" from the exception message as Module::toString is specified to return "module XXXX" or "unnamed module XXXX" so you'll end up duplication "module" in the message.

AlanBateman commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:
```diff
@@ -66,16 +62,25 @@
+ * with a memory segment. This can be done, using the {@link jdk.incubator.foreign.ResourceScope} abstraction, as shown below:
+ *
+ * <pre>{@code
+try (ResourceScope scope = ResourceScope.ofConfined()) {
```
The example might be out of date, I assume this should be newConfinedScope.

mlchung commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -135,24 +144,71 @@ static CLinker getInstance() {
      * @return the downcall method handle.
      * @throws IllegalArgumentException in the case of a method type and function descriptor mismatch.
+     * {@code --enable-native-access} is either absent, or does not mention the module name {@code M}, or
+     * {@code ALL-UNNAMED} in case {@code M} is an unnamed module.
```
line 147-148 are leftover from ICE and they should be removed.

mlchung commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java:
```diff
@@ -135,24 +144,71 @@ static CLinker getInstance() {
+     * @throws IllegalStateException if {@code scope} has been already closed, or if access occurs from a thread other
+     * than the thread owning {@code scope}.
+     * {@code --enable-native-access} is either absent, or does not mention the module name {@code M}, or
+     * {@code ALL-UNNAMED} in case {@code M} is an unnamed module.
```
line 208-209 are also left-over from the last patch.

mlchung commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractCLinker.java:
```diff
@@ -0,0 +1,60 @@
+public abstract class AbstractCLinker implements CLinker {
+    @CallerSensitive
+    public final MethodHandle downcallHandle(Addressable symbol, MethodType type, FunctionDescriptor function) {
+        Reflection.ensureNativeAccess(Reflection.getCallerClass());
```
`downcallHandle` method is no longer caller-sensitive.  `@CallerSensitive` and `Reflection.ensureNativeAccess` are no longer needed.    Same for the 4-arg `downcallHandle` method.

mlchung commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/LibrariesHelper.java:
```diff
@@ -29,27 +29,29 @@
 import jdk.internal.loader.NativeLibrary;
-import jdk.internal.ref.CleanerFactory;
+import jdk.internal.reflect.CallerSensitive;
+import jdk.internal.reflect.Reflection;
```
I think these 2 imports are no longer used.

mlchung commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/AArch64Linker.java:
```diff
@@ -25,28 +25,28 @@
 import jdk.internal.foreign.abi.SharedUtils;
 import jdk.internal.foreign.abi.UpcallStubs;
+import jdk.internal.reflect.CallerSensitive;
+import jdk.internal.reflect.Reflection;
```
These 2 imports are no longer needed.  Same for other CLinker implementation classes.

mlchung reviewed (approved):
The caller-sensitive cleanup change looks good.

mcimadamore commented:
Thanks for the review @mlchung - hopefully looks better now!

iwanowww reviewed (approved):
Overall, hotspot and java.lang.invoke changes look good.

One comment though on naming: there are multiple references to Panama in the code. Moreover, it is used inconsistently (`is_panama_entry_frame()` vs `EntryBlob` vs `OptimizedUpcallStub`). 
I suggest to get rid of any references to Panama and use `optimized`uniformly where appropriate (`is_optimized_entry_frame()`, `OptimizedEntryBlob`, `OptimizedUpcallStub`).

Some minor comments follow.

iwanowww commented on src/hotspot/cpu/x86/universalUpcallHandler_x86_64.cpp:
```diff
@@ -141,3 +149,677 @@ address ProgrammableUpcallHandler::generate_upcall_stub(jobject rec, jobject jab
+  }
+  __ block_comment("} preserve_callee_saved_regs ");
+
+  // TODO mxcsr
```
Anything left to do with mxcsr?

JornVernee replied:
I guess this slipped through with the initial PR.

JNI code loads the default value of mxcsr here and saves/restores the incoming native value: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/stubGenerator_x86_64.cpp#L265-L298

However, it does nothing on Windows, while the Windows x64 ABI also defines mxcsr to have a non-volatile portion (i.e. that should be saved and restored): https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-160#mxcsr

I think this made me a bit unsure about the need for saving and restoring mxcsr, and wanted to find an actual test case that causes a problem, rather than just copy-pasting the code. But, at this time I haven't gotten to that yet.

iwanowww replied:
Thanks for the pointers. It would be safer for now to align the behavior with the call stub.

mcimadamore commented on src/hotspot/share/runtime/sharedRuntime.hpp:
```diff
@@ -462,6 +462,11 @@ class SharedRuntime: AllStatic {
   static void    save_native_result(MacroAssembler *_masm, BasicType ret_type, int frame_slots);
   static void restore_native_result(MacroAssembler *_masm, BasicType ret_type, int frame_slots);
 
+  static void   move32_64(MacroAssembler* masm, VMRegPair src, VMRegPair dst);
```
There's already an issue for that:
https://bugs.openjdk.java.net/browse/JDK-8266257

I've upgraded the description to make it reflect the proposed move a bit more precisely.

iwanowww replied:
Got it. I was confused by the bug synopsis.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java:
```diff
@@ -194,14 +194,10 @@
  * Implementations of this interface are immutable, thread-safe and <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>.
  */
-public interface MemoryLayout extends Constable {
+public sealed interface MemoryLayout extends Constable permits AbstractLayout, SequenceLayout, GroupLayout, PaddingLayout, ValueLayout {
```
In principle we could just permit AbstractLayout and be done. In practice, the javadoc comes out better if we list all the concrete API interfaces which extend MemoryLayout, as the `permit` list will be displayed in the javadoc.

JornVernee replied:
At least the internal class name is hidden in the javadoc:
![image](https://user-images.githubusercontent.com/5962029/118983890-37042080-b97d-11eb-9ee0-ae5d5db5cd7e.png)

mcimadamore commented:
Latest javadoc:
http://cr.openjdk.java.net/~mcimadamore/JEP-412/v3/javadoc/jdk/incubator/foreign/package-summary.html
Latest specdiff:
http://cr.openjdk.java.net/~mcimadamore/JEP-412/v3/specdiff/overview-summary.html

mcimadamore commented:
/integrate
